### DISCUSSION TITLE
✔️Weekly Quests

### DISCUSSION THREAD

**Discussion by delgatojr** [zebscripts/AFK-Daily#63] [Ideas]
![image](https://user-images.githubusercontent.com/78246856/116646076-099ef680-a92c-11eb-8ae2-3d9442ad65b5.png)
Auto Ascend (20 points) should be easy.

![image](https://user-images.githubusercontent.com/78246856/116646201-5aaeea80-a92c-11eb-9789-18c357500e7d.png)

Then we are left with 4 quests for the 100 chest. Maybe we could do the last 2 quests by: clicking on those tabs, swiping all the way down (the guild quest will exclude this step), and clicking the cheapest item. To keep track, we can add a variable into the temp file that keeps track of when we bought these items. It will use `date +"%U"`, which outputs 
week number of year, with Sunday as first day of week (00..53). A function will run `date +"%U"` and if the value is different, then we buy, and re-set variable.

If we do this, then the time in tempFile (that we use for Mythic Trick) will need to be changed into a variable, as well change the corresponding functions.

**Comment by Zebiano** [comment]
I'd like to have separate functions for these actions, namely `autoAscendHeroes()`, `buyFromLabStore()` and `buyFromGuildStore()`, that can be toggled to `true` or `false` in the `config.sh`, defaulting them to `false`.

I'm not sure yet how much I like this. I love the idea of involving weekly quests, though the more I think of this, the more it feels like it's going to bloat the script. I can imagine it'll be harder than expected to implement these ideas, especially the `.tmp` one. But it's definitely doable!

We'd also have to decide what to buy from the stores. Especially the Guild Store will be difficult, as everything always changes and each player may have their own priorities. The only option we have is to buy from the top row, and it will most likely be a random one (for example the first tile). I do not like this at all though.

Regarding the Lab Store, users also have priorities... We can choose to buy whatever, and people then decide if they want to enable it, but I still don't like this option. I'd rather have the player be able to choose which reward the script should buy, but then that is freaking difficult to code and I wonder if it's worth it...

Ugh, I don't know. I don't know what to think of this. 😅

**Reply by kevingrillet** [reply]
Mmh, why not trying to move from `.tmp` to a true `.ini` file.

With readable dates:

- lastCampaign > `date +%Y%m%d` (`date +%Y%m%d -d "$(date) + 3 day"`)
- lastWeekly > `date -dlast-saturday +%Y%m%d`

```ini
[afk-daily]
lastCampaign=20210430
lastWeekly=20210424
```

Using things like that [lib](https://jonlabelle.com/snippets/view/shell/read-and-write-ini-files-in-bash) (my [script](https://github.com/kevingrillet/ShellUtils/blob/main/utils/utils_ini.sh) without `sed -r` or `awk` ~~I want to try developing my own this weekend, but this one is more than OK~~) or any of [this](https://stackoverflow.com/questions/6318809/how-do-i-grab-an-ini-value-within-a-shell-script) smaller scripts

<details>
<summary>do not use</summary>

**Broken on OSX: `sed: illegal option -- r`**

```shell
#!/bin/bash

# getFromIni <FILE> <SECTION> <PARAM>
getFromIni(){
    sed -nr "/^\[$2\]/ { :l /^$3[ ]*=/ { s/.*=[ ]*//; p; q;}; n; b l;}" "$1"
}

getFromIni .afkscript.ini afk-daily lastCampaign
```

Output: `20210424`

</details>

*PS: I'm not using Sunday because of this (today is Friday):*
```shell
$ date -dlast-friday +%Y%m%d
20210423

$ date +%Y%m%d
20210430
```

**Reply by delgatojr** [reply]
@Zebiano True. I did not think of people's priorities. That would definitely make it difficult to code. Oh well, at least auto ascend should be easy!

**Reply by delgatojr** [reply]
@kevingrillet I like that. It makes it look cleaner.

**Reply by Zebiano** [reply]
> Mmh, why not trying to move from .tmp to a true .ini file.

Now that I think about it, I should probably change `script.sh` to a `script.ini`...

**Reply by kevingrillet** [reply]
If we use an `.ini` file looking like [that](https://github.com/kevingrillet/ShellUtils/blob/main/examples/example_lighter.ini), we could use [this](https://github.com/kevingrillet/ShellUtils/blob/main/utils/utils_ini.sh) for:

- configFile="config.ini" (the actual config file will work perfectly fine, I think)
- tempFile=".afkscript.ini"
  ```ini
  # afk-daily
  lastCampaign=20210430
  lastWeekly=20210424
  ```

```shell
# getIniValueLighter <FILE> <PARAM>
# Output the param matching
# Warning: the ini file needs to don't have any [section]
# stdout result
getIniValueLighter() {
    if [ "$#" -ne 2 ] ; then echo "Usage: getIniValueLighter <FILE> <PARAM>" >&2; return 1; fi
    if [ -f "$1" ]; then                        # File exists
        . "$1"                                  # Add as source the .ini file
        echo "${!2}"                            # Echo the param (Substring expansion)
        # eval "echo \"\$$2\""                  # Alternative if the previous one does not work
    else
        echo "[ERROR] The file [$1] does not exist." >&2; return 1;
    fi
}
```

PS: We should check if the file is already included before including it again with a simple check on the variable.

*It's possible to add a default value if the file does not exist (and remove the error message ofc).*
*Look at my [repo](https://github.com/kevingrillet/ShellUtils/blob/main/utils/utils_ini.sh) to see which one you want to use.*

**Reply by Zebiano** [reply]
I also think there's no need to change the current `config.sh` file, it should work out if the box when changing it to `.ini`. 

Your repo is interesting! Has a lot of really nice small functions to help us out here, thanks! 

I do wonder though... Isn't there a better/native way to read the `.ini` file, besides using `read`? It feels kind of a downgrade then, as `.source` is (I imagine) way quicker and efficient, not to mention it auto-assigns variables.

**Reply by Zebiano** [reply]
Hm I've done some research and it doesn't look like you can... I'd rather keep the `script.sh` as it is then. It'll just bloat the script if we want to change it to `.ini` just because why not.

Maybe we should use a `.sh` file instead of a `.tmp` file since we're going to be storing variables there as well. Looks like the way to go there in my opinion. Suggestions are welcome!

**Reply by kevingrillet** [reply]
The fastest way to read a `.ini` file is to use this `sed` command, but it's not safe because it's using `sed -r` that does not exist everywhere:

```sh
# getFromIni <FILE> <SECTION> <PARAM>
getFromIni(){
    sed -nr "/^\[$2\]/ { :l /^$3[ ]*=/ { s/.*=[ ]*//; p; q;}; n; b l;}" "$1"
}
```

If you use `source` the extension is not important, we just need to be careful to not use `[section]` but just `# section`. Once it's included just `$param` et voila 😄

**Reply by Zebiano** [reply]
Yeah, I'd rather not have such a function.

**Reply by kevingrillet** [reply]
I did add it during my refacto, it's working at the moment.
